src/librustc/mir/mod.rs

```diff
@@ -225,13 +224,19 @@ impl<'tcx> Body<'tcx> {
     }
 
     #[inline]
-    pub fn predecessors(&self) -> MappedReadGuard<'_, IndexVec<BasicBlock, Vec<BasicBlock>>> {
-        self.cache.predecessors(self)
+    pub fn predecessors_ref(&self) -> &IndexVec<BasicBlock, Vec<BasicBlock>> {
+        self.cache.predecessors_ref()
     }
 
     #[inline]
-    pub fn predecessors_for(&self, bb: BasicBlock) -> MappedReadGuard<'_, Vec<BasicBlock>> {
-        MappedReadGuard::map(self.predecessors(), |p| &p[bb])
+    pub fn predecessors_mut(&mut self) -> &mut IndexVec<BasicBlock, Vec<BasicBlock>> {
+        // TODO(nashenas88) figure out a way to get rid of this clone
+        self.cache.predecessors_mut(&self.clone())
```

Comment:
I think we should just inline the `predecessors` field from the `Cache` into `Body` (as `predecessors_cache` I guess), thus eliminating the `cache` field entirely. Then you can move all the logic from `Cache::predecessors_mut` here and remove the `Cache` type, too.